### PR TITLE
feat: add standalone setup-gateway.sh script for maas-default-gateway

### PR DESCRIPTION
**Pull request by somya-bhatnagar**
## Summary

Extracts gateway setup logic from `deploy.sh` into a standalone `scripts/setup-gateway.sh` script that supports two deployment modes for different cluster environments.

**Deployment Modes:**
- **route mode** (default): LoadBalancer Service with cluster router TLS - best for ROSA, OSD, cloud clusters
- **clusterip mode**: ClusterIP Service + OpenShift Route with service-ca TLS - best for on-prem, disconnected, bare-metal clusters

**Features:**
- Auto-detects cluster domain and TLS certificate (four-level fallback)
- Validates gateway spec matches requested mode before early exit (prevents mode mismatch issues)
- Supports `DISCONNECTED=true` flag for air-gapped clusters (no GitHub fallback)
- Idempotent with comprehensive annotation and spec validation
- Environment variable overrides: `INGRESS_MODE`, `CLUSTER_DOMAIN`, `CERT_NAME`, `DISCONNECTED`, `DRY_RUN`

## Changes

- **New:** `scripts/setup-gateway.sh` - Standalone gateway setup script (547 lines)
- **Modified:** `scripts/deploy.sh` - Removed deprecated inline functions, calls new script
- **Modified:** `scripts/README.md` - Added comprehensive documentation for `setup-gateway.sh` and `setup-database.sh`
- **Modified:** `docs/content/install/maas-setup.md` - Updated to script-based approach with tested environments

## Live Cluster Testing

Validated on 3 cluster types across different cloud providers and network environments:

### ✅ ROSA Cluster (AWS us-east-1)
- **Route mode:** Gateway Programmed, TLS from IngressController (primary detection method)
- **ClusterIP mode:** Gateway Programmed, service-ca provisioned TLS (~10s), Route Admitted
- **Idempotency:** Clean re-runs on existing resources

### ✅ AWS Cluster (AWS us-west-1)  
- **Route mode:** Gateway Programmed, TLS from router deployment (fallback 1)
- **ClusterIP mode:** Gateway Programmed, service-ca provisioned TLS (~12s), Route Admitted
- **Mode switching:** Validated spec validation prevents clusterip mode from skipping on route-mode gateway

### ✅ Disconnected Cluster (Air-gapped)
- **ClusterIP mode:** Gateway Programmed using local manifests
- **DISCONNECTED=true:** Confirmed GitHub fallback is blocked, fails fast on missing local manifests
- **TLS detection:** Used router deployment certificate (fallback 1)

### Test Coverage
- ✅ TLS auto-detection: All 4 fallback levels validated (IngressController → router deployment → known secrets → self-signed)
- ✅ Idempotency: 6 scenarios tested (existing GatewayClass, Gateway, ConfigMap, Route, annotations, spec)
- ✅ Mode validation: Prevents mode mismatch by checking `infrastructure.parametersRef`
- ✅ Error handling: Validated failure paths (missing tools, bad mode, network issues)

## Risk Analysis

**Risk rating:** 2

**Why:**
- New script extraction reduces complexity in `deploy.sh` (positive)
- Tested on 3 cluster types with comprehensive validation (reduces risk)
- Changes gateway setup flow that is part of deploy path but not all edge cases covered by Prow smoke test
- Mode selection (`INGRESS_MODE=clusterip`) and DISCONNECTED flag paths are new, not in current CI
- Impact limited to gateway setup phase; does not affect existing gateways or running workloads

## References

- Implements: [RHOAIENG-63963](https://issues.redhat.com/browse/RHOAIENG-63963)
- Pattern: Follows `setup-database.sh` standalone script pattern
- Documentation: Comprehensive usage in `scripts/README.md` and `docs/content/install/maas-setup.md`

---

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Documentation**
  * Simplified Gateway setup guide with a script-driven workflow, mode-specific commands, TLS guidance, and explicit verification steps; README documents mode options and environment variables (DRY_RUN, DISCONNECTED, overrides).
  * Added follow-up step to run an Authorino TLS setup with namespace guidance.

* **New Features**
  * Automated gateway provisioning supporting route vs clusterip modes, automatic TLS resolution with fallbacks, and verification/waiting for Gateway readiness.

* **Refactor**
  * Deployment flow now delegates Gateway provisioning to the centralized setup script.

### CODE REVIEW

**Comment by somya-bhatnagar** [comment]
@coderabbitai review

File: scripts/README.md [hunk 1]
```diff
@@ -196,6 +196,89 @@ AUTHORINO_NAMESPACE=rh-connectivity-link ./scripts/setup-authorino-tls.sh
...
+- `POSTGRES_USER` - Database user (default: maas)
+- `POSTGRES_DB` - Database name (default: maas)
+- `POSTGRES_PASSWORD` - Database password (default: auto-generated)
+
+**Production Use:**
+This script deploys POC-grade PostgreSQL with ephemeral storage. For production:
+```bash
+# Use deploy.sh with external database connection string
+./scripts/deploy.sh --postgres-connection 'postgresql://user:pass@host:5432/db?sslmode=require'
+```
```
[the remark this answers is not in the document]
Comment (somya-bhatnagar) [reply scripts/README.md]: Done

File: docs/content/install/maas-setup.md [hunk 2]
```diff
@@ -47,10 +52,14 @@ postgresql://USERNAME:PASSWORD@HOSTNAME:PORT/DATABASE?sslmode=require
...
-    If you use `scripts/deploy.sh`, you can supply your own PostgreSQL connection string with the `--postgres-connection` flag. This skips the built-in POC PostgreSQL deployment and creates the `maas-db-config` Secret automatically:
+    If you use `scripts/deploy.sh`, you can supply your own PostgreSQL connection string with the `--postgres-connection` flag (or set `POSTGRES_CONNECTION` in the environment). This skips the built-in POC PostgreSQL deployment and creates the `maas-db-config` Secret automatically.
+
+    Avoid embedding credentials in the command line; export the connection string, pass it to the flag, then unset it:
 
     ```bash
-    ./scripts/deploy.sh --postgres-connection 'postgresql://username:password@hostname:5432/database?sslmode=require'
+    export POSTGRES_CONNECTION='postgresql://username:password@hostname:5432/database?sslmode=require'
+    ./scripts/deploy.sh --postgres-connection "$POSTGRES_CONNECTION"
+    unset POSTGRES_CONNECTION
```
Comment (chaitanya1731) [inline docs/content/install/maas-setup.md]: These are DB related changes in the doc, lets not include these in this PR. Only scope this for Gateway Setup.

File: docs/content/install/maas-setup.md [hunk 3]
```diff
@@ -64,78 +73,33 @@ postgresql://USERNAME:PASSWORD@HOSTNAME:PORT/DATABASE?sslmode=require
...
-       certificateRefs:
-       - group: ""
-         kind: Secret
-         name: ${CERT_NAME}
-       mode: Terminate
-EOF
+```bash
+kubectl wait --for=condition=Programmed gateway/maas-default-gateway -n openshift-ingress --timeout=120s
+# ClusterIP mode only:
+kubectl get route maas-gateway-route -n openshift-ingress
```
Comment (chaitanya1731) [inline docs/content/install/maas-setup.md:92]: Can you include the expected output here to verify readiness?

File: scripts/README.md [hunk 4]
```diff
@@ -46,6 +46,7 @@ Automated deployment script for OpenShift clusters supporting both operator-base
 - `--operator-catalog <image>` - Custom operator catalog image for PR testing
 - `--operator-image <image>` - Custom operator image for PR testing
 - `--channel <channel>` - Operator channel override (default: fast-3 for ODH, stable-3.x for RHOAI)
+- `--postgres-connection <connection-string>` - External PostgreSQL URL (also reads `POSTGRES_CONNECTION` env var). Avoid putting credentials on the command line; export the variable and pass `"$POSTGRES_CONNECTION"` instead.
```
Comment (chaitanya1731) [inline scripts/README.md]: again same as DB comment above..

File: scripts/README.md [hunk 5]
```diff
@@ -196,6 +197,48 @@ AUTHORINO_NAMESPACE=rh-connectivity-link ./scripts/setup-authorino-tls.sh
...
+- `DISCONNECTED` - Disable GitHub manifest fallback (`true`/`false`, default: false)
+- `DRY_RUN` - Preview changes without applying (`true`/`false`, default: false)
+- `MAAS_MANIFEST_REF` - Git tag or commit SHA for remote kustomize fallback (defaults to current repo `HEAD` when run from a clone; required when fetching without a local tree)
+
+**TLS Auto-Detection (Route Mode):**
+The script automatically detects TLS certificates using this priority:
+1. IngressController certificate
+2. Router deployment certificate
+3. Known certificate secrets
+4. Self-signed certificate (fallback)
```
Comment (chaitanya1731) [inline scripts/README.md]: Not sure if we need to provide detailed info? can we make it concise to only keep instructions about usage.

File: scripts/setup-gateway.sh [hunk 6]
```diff
@@ -0,0 +1,620 @@
...
+DRY_RUN="${DRY_RUN:-false}"
+MAAS_MANIFEST_REF="${MAAS_MANIFEST_REF:-}"
+
+# Constants
+GATEWAY_NAMESPACE="openshift-ingress"
+GATEWAY_NAME="maas-default-gateway"
+GATEWAYCLASS_NAME="openshift-default"
+GW_OPTIONS_CONFIGMAP="gw-options"
+GATEWAY_ROUTE_NAME="maas-gateway-route"
+SERVICE_CA_SECRET="maas-gw-service-tls"
```
Comment (chaitanya1731) [inline scripts/setup-gateway.sh:68]: I think we are creating the secret name from the data-science default gateway? in the implementation of @dmytro-zaharnytskyi 's PR for ODH. this may create mismatch.

File: docs/content/install/maas-setup.md (hunk 2 above)
Comment (somya-bhatnagar) [reply docs/content/install/maas-setup.md]: Agree on this, let me clear off this.

File: scripts/setup-gateway.sh (hunk 6 above)
Comment (somya-bhatnagar) [reply scripts/setup-gateway.sh:68]: This PR maintains the existing naming convention. The alignment with ODH operator's "maas-gateway-tls" can be addressed separately in a follow-up PR that updates all references across the codebase, not mixed into this gateway script extraction PR. your thoughts ?

image

File: docs/content/install/maas-setup.md [hunk 7]
```diff
@@ -64,79 +64,48 @@ postgresql://USERNAME:PASSWORD@HOSTNAME:PORT/DATABASE?sslmode=require
...
+Expected output:
+```
+NAME                  HOST/PORT                           PATH   SERVICES                           PORT   TERMINATION   WILDCARD
+maas-gateway-route    maas.apps.example.cluster.com              maas-default-gateway-openshift-default   443    reencrypt     None
 ```
 
+For ClusterIP architecture and troubleshooting, see [Gateway patterns — ClusterIP + Route](../configuration-and-management/gateway-patterns.md#clusterip-gateway-with-openshift-route-re-encrypt).
+
+!!! note "Required annotations"
+    The script applies `opendatahub.io/managed: "false"` and `security.opendatahub.io/authorino-tls-bootstrap: "true"` on the Gateway. See [TLS Configuration](../configuration-and-management/tls-configuration.md) for custom certificates and Authorino integration. The `authorino-tls-bootstrap` annotation is interim until [CONNLINK-528](https://issues.redhat.com/browse/CONNLINK-528).
```
Comment (chaitanya1731) [inline docs/content/install/maas-setup.md]:
```suggestion
    The script applies `opendatahub.io/managed: "false"` and `security.opendatahub.io/authorino-tls-bootstrap: "true"` on the Gateway. See [TLS Configuration](../configuration-and-management/tls-configuration.md) for custom certificates and Authorino integration.
```

File: scripts/setup-gateway.sh [hunk 8]
```diff
@@ -0,0 +1,620 @@
...
+
+  kubectl apply -f - <<EOF
+apiVersion: route.openshift.io/v1
+kind: Route
+metadata:
+  name: ${GATEWAY_ROUTE_NAME}
+  namespace: ${GATEWAY_NAMESPACE}
+  annotations:
+    # Tells the OpenShift Router to trust the service-ca certificate on the backend.
+    router.openshift.io/service-ca-certificate: "true"
```
[the remark this answers is not in the document]
Comment (dmytro-zaharnytskyi) [reply scripts/setup-gateway.sh]: Agree with coderabbit here. Is there any reason why this approach was selected? On clusters where the router doesn't implicitly trust service-ca, the reencrypt handshake will probably fail silently.

**Review by dmytro-zaharnytskyi — COMMENTED** [review]
Went through the PR, no major concerns. Only one code rabbit comment that I agree with. Ping me after referencing (or disregarding with some explanation) it and I will approve 🤗

**Review by dmytro-zaharnytskyi — APPROVED** [review]
lgtm

**Review by chaitanya1731 — APPROVED** [review]
Thanks Somya.. 

/lgtm
/approve
